{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulate sequences from a lineage tracing experiment \n",
    "\n",
    "Simulate $N$ sequences resulting from Cas9 cutting over $T$ timesteps to be used for testing alignment strategies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 599,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "\n",
    "T = 3\n",
    "N = 100000"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in the reference sequence to be mutated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 600,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_file = \"/home/mattjones/PCT48_seq.fasta\"\n",
    "ref_seq = \"\"\n",
    "with open(ref_file, \"r\") as handle:\n",
    "    ref_seq = str(list(SeqIO.parse(handle, \"fasta\"))[0].seq)\n",
    "    \n",
    "extension = \"TAAACCCGCTGATCAGCCTCGACTGTGCCTTCTAGTTGCCAGCCATCTGTTGTTTGCCCCTCCCCCGTGCCTTCCTTGACCCTGGAAGGTGCCACTCCCACTGTCCTTTCCTAATAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\"\n",
    "\n",
    "# These are the locations for the cut sites in the reference sequence.\n",
    "cut_site_locs = {-1: 94, 0: 112, 1: 166, 2: 220}\n",
    "REF_SEQ_LEN = 296"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 605,
   "metadata": {},
   "outputs": [],
   "source": [
    "class TargetSiteSequence:\n",
    "    \n",
    "    def __init__(self, seq, ident, needle=True):\n",
    "        \n",
    "        self.seq = seq\n",
    "        self.add_intBC()\n",
    "        self.id = ident\n",
    "        \n",
    "        # Boolean values indicating whether or not the cute sites have been cut yet\n",
    "        self.is_cut = np.array([False, False, False])\n",
    "        self.cut_sites = {-1: 94, 0: 112, 1: 166, 2: 220}\n",
    "        self.allele_cigar = [\"\"]*3\n",
    "        \n",
    "        self.needle = needle\n",
    "            \n",
    "    def cut_site(self):\n",
    "        \"\"\" \n",
    "        Cut sites in sequence with N_CAS9 cas9 proteins.\n",
    "        \"\"\"\n",
    "        \n",
    "        # all cut sites have been cut!\n",
    "        if all(self.is_cut):\n",
    "            return\n",
    "        \n",
    "        nts = [\"A\", \"C\", \"T\", \"G\"]\n",
    "        \n",
    "        bool_to_cigar = {True: \"D\", False: \"I\"}\n",
    "        \n",
    "        n_cas9 = draw_ncas9()\n",
    "        \n",
    "        # no cutting if no cas9 come to the cut site\n",
    "        if n_cas9 == 0:\n",
    "            return\n",
    "        \n",
    "        \n",
    "        sites = self.draw_sites(n_cas9)\n",
    "\n",
    "        if len(sites) == 1:\n",
    "            s = sites[0]\n",
    "            # Introduce random indel in this particular site\n",
    "            \n",
    "            # choose location in site\n",
    "            loc = pick_site(cut_site_locs[s])\n",
    "            \n",
    "            # choose length of indel\n",
    "            indel_length = draw_indel_length()\n",
    "            \n",
    "            # flip coin to be insertion or deletion\n",
    "            is_del = (0.75 > np.random.uniform())\n",
    "            \n",
    "            # Add indel to self.allele_cigar         \n",
    "            self.allele_cigar[s] = str(indel_length) + bool_to_cigar[is_del]\n",
    "            \n",
    "            # Add indel to self.seq\n",
    "            if is_del:\n",
    "                self.seq = (self.seq[:loc] + self.seq[(loc+indel_length):])\n",
    "                \n",
    "            else:\n",
    "                inserted_nt = \"\".join(np.random.choice(nts, indel_length))\n",
    "                self.seq = (self.seq[:loc] + inserted_nt + self.seq[(loc):])\n",
    "                \n",
    "            self.cut_sites[s] = loc\n",
    "            self.is_cut[s] = True\n",
    "                \n",
    "        elif len(sites) == 2:\n",
    "            # if there are two sites being cut simultaenously, we need to remove the segment between cut sites\n",
    "            s1, s2 = min(sites[0], sites[1]), max(sites[0], sites[1])\n",
    "            loc1, loc2 = pick_site(cut_site_locs[s1]), pick_site(cut_site_locs[s2])\n",
    "            \n",
    "            self.allele_cigar[s1] = str(loc2 - loc1) + \"D\"\n",
    "            self.allele_cigar[s2] = \"Resected\"\n",
    "            \n",
    "            if s1 == 0 and s2 == 2:\n",
    "                self.allele_cigar[1] = \"Resected\"\n",
    "                self.is_cut[1] = True \n",
    "                \n",
    "                \n",
    "            self.seq = self.seq[:loc1] + self.seq[(loc2):]\n",
    "            self.cut_sites[s1], self.cut_sites[s2] = loc1, loc2\n",
    "            self.is_cut[s1], self.is_cut[s2] = True, True\n",
    "            \n",
    "        else:\n",
    "            s1, s2 = 0, 2\n",
    "            loc1, loc2 = pick_site(cut_site_locs[s1]), pick_site(cut_site_locs[s2])\n",
    "            \n",
    "            self.allele_cigar[s1] = str(loc2 - loc1) + \"D\"\n",
    "            self.allele_cigar[s2] = \"Resected\"\n",
    "            \n",
    "            self.allele_cigar[1] = \"Resected\"\n",
    "            \n",
    "            self.seq = self.seq[:loc1] + self.seq[(loc2):]\n",
    "            \n",
    "            self.cut_sites[0], self.cut_sites[2] = loc1, loc2\n",
    "            self.is_cut[0], self.is_cut[1], self.is_cut[2] = True, True, True\n",
    "\n",
    "        \n",
    "        \n",
    "    def add_intBC(self):\n",
    "        \n",
    "        alphabet = [\"A\", \"C\", \"G\", \"T\"]\n",
    "        bc = \"\".join(np.random.choice(alphabet, 14))\n",
    "        \n",
    "        self.seq = self.seq.replace(\"N\"*14, bc)\n",
    "        \n",
    "    def draw_sites(self, n_cas9):\n",
    "        \n",
    "        unavail_sites = np.where(self.is_cut == True)[0]\n",
    "        \n",
    "        probs = np.ones(len(self.is_cut)) / (3 - len(unavail_sites))\n",
    "        probs[unavail_sites] = 0\n",
    "        \n",
    "        sites = np.random.choice(3, size=min(n_cas9, 3 - len(unavail_sites)), p=probs, replace=False)\n",
    "        \n",
    "        return sites\n",
    "    \n",
    "    def add_mutations(self, l=5):\n",
    "        \n",
    "        nt = [\"A\", \"C\", \"G\", \"T\"]\n",
    "        n_muts = np.random.poisson(5)\n",
    "        \n",
    "        for i in range(n_muts):\n",
    "            loc = np.random.choice(len(self.seq))\n",
    "            c_nt = self.seq[loc]\n",
    "            n_nt = np.random.choice(nt)\n",
    "            \n",
    "            while c_nt == n_nt:\n",
    "                n_nt = np.random.choice(nt)\n",
    "            \n",
    "            self.seq = self.seq[:loc] + n_nt + self.seq[loc+1:]\n",
    "    \n",
    "def draw_indel_length(N=0.5, P=0.1):        \n",
    "    \n",
    "    return min(30, max(1, np.random.negative_binomial(N, P)))\n",
    "        \n",
    "def draw_ncas9(l = 0.4):\n",
    "    \n",
    "    return min(3, np.random.poisson(lam = l))\n",
    "\n",
    "def pick_site(loc):\n",
    "    \"\"\"\n",
    "    Draw a random site to begin a cut at, centered at loc.\n",
    "    \"\"\"\n",
    "    \n",
    "    offset = max(-14, min(4, 10 - np.random.poisson(10)))\n",
    "    \n",
    "    return loc + offset\n",
    "\n",
    "def create_cigar_string(cigar_list, cut_sites):\n",
    "    \n",
    "    \n",
    "    cs = [str(cut_site_locs[-1]) + \"M\"]\n",
    "    for i in range(3):\n",
    "        \n",
    "        if cigar_list[i] == \"Resected\":\n",
    "            continue\n",
    "        \n",
    "        if cigar_list[i] == \"\":\n",
    "            \n",
    "            match_len = cut_sites[i] - cut_sites[i-1]\n",
    "            \n",
    "            if i != 0 and cigar_list[i-1] != \"Resected\":\n",
    "                if cs[len(cs) - 1][-1] == \"D\":\n",
    "                    match_len -= int(cs[len(cs)-1][:-1])\n",
    "\n",
    "            if cs[len(cs) - 1][-1] == \"M\":\n",
    "\n",
    "                match_len = match_len + int(cs[len(cs) -1][:-1])\n",
    "                cs[len(cs) - 1] = str(match_len) + \"M\"\n",
    "\n",
    "\n",
    "            else:\n",
    "                cs += [str(match_len) + \"M\"]\n",
    "            \n",
    "            \n",
    "        else:\n",
    "            match_len = cut_sites[i] - cut_sites[i-1]\n",
    "            \n",
    "            if cs[len(cs) - 1][-1] != \"M\":\n",
    "                if i != 0 and cigar_list[i-1] != \"Resected\":\n",
    "                    if cs[len(cs) - 1][-1] == \"D\":\n",
    "                        match_len -= int(cs[len(cs)-1][:-1])\n",
    "                    cs += [str(match_len) + \"M\"] + [cigar_list[i]]\n",
    "                else:\n",
    "                    cs += [str(match_len) + \"M\"] + [cigar_list[i]]\n",
    "                \n",
    "                \n",
    "            if cs[len(cs) - 1][-1] == \"M\":\n",
    "                match_len = match_len + int(cs[len(cs) -1][:-1])\n",
    "                cs[len(cs) - 1] = str(match_len) + \"M\"\n",
    "                cs += [cigar_list[i]]\n",
    "    \n",
    "    final_match_len = REF_SEQ_LEN - cut_sites[2]\n",
    "    \n",
    "    if cigar_list[2] != \"Resected\":\n",
    "        if cs[len(cs) - 1][-1] == \"D\":\n",
    "            final_match_len -= int(cs[len(cs) - 1][:-1])\n",
    "            \n",
    "    \n",
    "    if cs[len(cs) - 1][-1] == \"M\":\n",
    "        final_match_len = final_match_len + int(cs[len(cs) -1][:-1])\n",
    "        cs[len(cs) - 1] = str(final_match_len) + \"M\"\n",
    "        \n",
    "    else:\n",
    "        cs += [str(final_match_len) + \"M\"]\n",
    "    \n",
    "        \n",
    "    \n",
    "    return \"\".join(cs), cs\n",
    "\n",
    "            \n",
    "def test_cigar_caller(cigar):\n",
    "    s = 0\n",
    "    for c in cigar:\n",
    "        if c[-1] != \"I\":\n",
    "            s += int(c[:-1])\n",
    "    \n",
    "    return s == REF_SEQ_LEN\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now simulate sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 606,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_file = open(\"simulated_target_sites.needle.txt\", \"w\")\n",
    "fa_out = open(\"simulated_target_sites.needle.fa\", \"w\")\n",
    "\n",
    "out_file.write(\"ID\\tSequence\\tCigar\\n\")\n",
    "\n",
    "for n in range(N):\n",
    "    \n",
    "    seq = TargetSiteSequence(ref_seq, \"seq_\" + str(n))\n",
    "    for t in range(min(5, np.random.poisson(T))):\n",
    "        seq.cut_site()\n",
    "        \n",
    "    cs, cs_list = create_cigar_string(seq.allele_cigar, seq.cut_sites)\n",
    "    \n",
    "    assert test_cigar_caller(cs_list) == True\n",
    "    \n",
    "    seq.add_mutations()\n",
    "    \n",
    "    out_file.write(seq.id + \"\\t\" + seq.seq + '\\t' + cs + \"\\n\")\n",
    "    fa_out.write(\">\" + seq.id + \"\\n\" + seq.seq + '\\n')\n",
    "    \n",
    "out_file.close()\n",
    "    \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
